### PR TITLE
Move Broadcast of full blocks before processing block

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes:
- Broadcast full blocks as soon as they are validated, without processing as per spec: https://github.com/ethereum/devp2p/blob/master/caps/eth.md#block-propagation

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by marcindsobczak — APPROVED** [review]

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs [hunk 1]
```diff
@@ -181,34 +181,41 @@ public void AddNewBlock(Block block, ISyncPeer nodeWhoSentTheBlock)
...
-            lock (_recentlySuggested)
+            if (isBlockBeforeTheSyncPivot || isBlockOlderThanMaxReorgAllows)
             {
-                if (_recentlySuggested.Get(block.Hash) != null) return;
-                _recentlySuggested.Set(block.Hash, _dummyValue);
+                return;
             }
 
-            if (_specProvider.TerminalTotalDifficulty != null && block.TotalDifficulty >= _specProvider.TerminalTotalDifficulty)
+            if (_recentlySuggested.Set(block.Hash))
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/SyncServer.cs:190]: we removed lock(_recentlySuggested). Is that okay?

Comment (MarekM25) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:190]: and we changed a bit logic here

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs [hunk 2]
```diff
@@ -181,34 +181,41 @@ public void AddNewBlock(Block block, ISyncPeer nodeWhoSentTheBlock)
...
-            ValidateSeal(block, nodeWhoSentTheBlock);
-            if ((_syncModeSelector.Current & (SyncMode.FastSync | SyncMode.StateNodes)) == SyncMode.None
-                || (_syncModeSelector.Current & SyncMode.Full) != SyncMode.None)
-            {
-                LogBlockAuthorNicely(block, nodeWhoSentTheBlock);
-                SyncBlock(block, nodeWhoSentTheBlock);
+                ValidateSeal(block, nodeWhoSentTheBlock);
+                SyncMode syncMode = _syncModeSelector.Current;
+                bool notInFastSyncNorStateSync = (syncMode & (SyncMode.FastSync | SyncMode.StateNodes)) == SyncMode.None;
+                bool inFullSyncOrWaitingForBlocks = (syncMode & (SyncMode.Full | SyncMode.WaitingForBlock)) != SyncMode.None;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/SyncServer.cs:200]: oh so before your changes if we were in WaitingForBlock we won't accept gossip blocks?

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs [hunk 3]
```diff
@@ -181,34 +181,41 @@ public void AddNewBlock(Block block, ISyncPeer nodeWhoSentTheBlock)
...
+                if (notInFastSyncNorStateSync || inFullSyncOrWaitingForBlocks)
+                {
+                    LogBlockAuthorNicely(block, nodeWhoSentTheBlock);
+                    SyncBlock(block, nodeWhoSentTheBlock);
+                }
+
+                // we might null total difficulty for a block in a block tree as we don't trust the message
+                Block blockToBroadCast = block.TotalDifficulty is null
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/SyncServer.cs:208]: so we're rebroadcasting block with potentially wrong difficulty. What if in SyncBlock headerValidator rejected block? Will we broadcast invalid block?

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs [hunk 4]
```diff
@@ -249,34 +254,70 @@ private void SyncBlock(Block block, ISyncPeer? syncPeer)
...
                 {
-                    string message = $"Peer {syncPeer?.Node:c} sent an invalid block.";
+                    string message = $"Peer {nodeWhoSentTheBlock?.Node:c} sent an invalid block.";
                     if (_logger.IsDebug) _logger.Debug(message);
-                    lock (_recentlySuggested)
-                    {
-                        _recentlySuggested.Delete(block.Hash!);
-                    }
-
+                    _recentlySuggested.Delete(block.Hash!);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/SyncServer.cs:259]: oh maybe lock is no longer needed, but why?

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:190]: yes, we don't use Get and Set separately. The result of Set should be enough for us and Set itself is thread-safe so we don't need to have an outside lock.

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:200]: Not sure but it didn't look right to me.

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs [hunk 5]
```diff
@@ -249,34 +254,70 @@ private void SyncBlock(Block block, ISyncPeer? syncPeer)
...
                 if (_logger.IsTrace) _logger.Trace($"SyncServer block {block.ToString(Block.Format.FullHashAndNumber)}, SuggestBlock result: {result}.");
             }
             else
             {
-                if (_logger.IsDebug) _logger.Debug($"Peer {syncPeer} sent block with unknown parent {block}, best suggested {_blockTree.BestSuggestedHeader}.");
+                if (_logger.IsDebug) _logger.Debug($"Peer {nodeWhoSentTheBlock} sent block with unknown parent {block}, best suggested {_blockTree.BestSuggestedHeader}.");
             }
         }
 
+        private void BroadcastBlock(Block block, bool allowHashes, ISyncPeer? nodeWhoSentTheBlock = null)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/SyncServer.cs:279]: so as I understand we will broadcast the block two times.
1. After NewBlockMessage
2. After we marked it as a head

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:208]: If `ValidateSeal` or `SyncBlock` reject the block then they throw an exception, so we don't rebroadcast the block. Let me think if I can refactor it to make it clearer.

File: src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs [hunk 6]
```diff
@@ -30,22 +30,23 @@ namespace Nethermind.Consensus.Ethash
...
+        private readonly ICryptoRandom _cryptoRandom;
+        private readonly IEthash _ethash;
+        private readonly ITimestamper _timestamper;
+        private readonly ILogger _logger;
 
-        private ICache<Keccak, bool> _sealCache = new LruCache<Keccak, bool>(2048, 2048, "ethash seals");
+        private readonly ICache<Keccak, bool> _sealCache = new LruCache<Keccak, bool>(2048, 2048, "ethash seals");
         private const int SealValidationIntervalConstantComponent = 1024;
-        private int sealValidationInterval = SealValidationIntervalConstantComponent;
+        private const long AllowedFutureBlockTimeSeconds = 15;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs:41]: Did we find this value in geth?

File: src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs [hunk 7]
```diff
@@ -90,17 +91,23 @@ public bool ValidateParams(BlockHeader parent, BlockHeader header)
...
             bool isDifficultyCorrect = difficulty == header.Difficulty;
             if (!isDifficultyCorrect)
             {
-                _logger.Warn(
-                    $"Invalid block header ({header.ToString(BlockHeader.Format.Full)}) - incorrect difficulty {header.Difficulty} instead of {difficulty}");
+                if (_logger.IsWarn) _logger.Warn($"Invalid block header ({header.ToString(BlockHeader.Format.Full)}) - incorrect difficulty {header.Difficulty} instead of {difficulty}");
+                return false;
+            }
+
+            ulong unixTimeSeconds = _timestamper.UnixTime.Seconds;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs:106]: so we started validating blocks using clock, right?

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 4 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:259]: LruKeyCache is thread-safe by design

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 5 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:279]: Yes, this is a compromise.
1. After NewBlockMessage we will broadcast full block to sqrt(peers)
2. After becoming head we will broadcast full block to sqrt(peers) and hash to the rest.

In theory we potentially shouldn't broadcast full blocks again after head, but we would need somehow to store this information. Geth stores in a block information about the peer it came from. We could do that too.

But I didn't found any major drawbacks of broadcasting it again to this fairly small set of peers, so... I ignored it for now.

File: src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs (hunk 6 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs:41]: yes

File: src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs (hunk 7 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus.Ethash/EthashSealValidator.cs:106]: only in PoW

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 2 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:200]: yup for me too

File: src/Nethermind/Nethermind.Synchronization/SyncServer.cs (hunk 3 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs:208]: but wait, imagine that we don't have the parent of this block. In this case we will rebrodcast it. Shouldn't we call BroadcastBlock in SyncBlock method?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SyncServer.cs]: Refactored it

**Comment by LukaszRozmej** [comment]
Looks to work fine on my node

**Pull request merged**